File: algo_05_dynamic_plan/dp_2_subsequence/L152-m.cpp
//
// Created by LiHeng on 2020/6/3.
//

#include <iostream>
#include <vector>
using namespace std;
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int sz = nums.size();
        int tmp_max = nums[0];
        int tmp_min = nums[0];
        int final_max = nums[0];
        for(int i=1; i<sz; ++i){
            //考虑到负数的情况，每一步都要保存最大正值和最小负值
            //先把上一步的最大值保存下来，防止下面被修改掉
            int tmp = tmp_max;
            tmp_max = max(max(tmp_max*nums[i], tmp_min*nums[i]), nums[i]);
            tmp_min = min(min(tmp*nums[i], tmp_min*nums[i]), nums[i]);
            final_max = max(final_max, tmp_max);
        }
        return final_max;
    }
};

File: python-algorithm/algo_05_dynamic_plan/dp_2_subsequence/L152-m.py
#--coding:utf-8--

class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        tmp_max = nums[0]
        tmp_min = nums[0]
        final_max = nums[0]
        for i in range(1, len(nums)):
            # 标记上一步的最大值，防止被下面覆盖丢失
            tmp = tmp_max
            tmp_max = max(nums[i], max(tmp_max*nums[i], tmp_min*nums[i]))
            tmp_min = min(nums[i], min(tmp*nums[i], tmp_min*nums[i]))
            final_max = max(final_max, tmp_max)
        return final_max
